datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java

```diff
@@ -478,7 +478,13 @@ public ActionResult<Void> stop(@PathKeysParam PathKeys pathKeys,
           // this check helps in preventing any datastream from being stuck in STOPPING state indefinitely
           LOG.warn("Datastream {} is already in {} state. Notifying leader to initiate transition", d,
               d.getStatus());
-          _store.updateDatastream(d.getName(), d, true);
+          if (force) {
+            _store.deleteAssignmentTokens(d.getName());
+            d.setStatus(DatastreamStatus.STOPPED);
+            _store.updateDatastream(d.getName(), d, false);
+          } else {
+            _store.updateDatastream(d.getName(), d, true);
```

Comment:
The purpose of this call is to recover the setup and to do cleanup of pending tokens in case leader failed to do so. This is a last resort call and this should be used sparingly.

Reply:
The only issue i have is since you [fetch the stopping datastreams from the _liveTaskMap](https://github.com/linkedin/brooklin/blob/091ce5d1e60106c72cf015363b997e2744d5e05e/datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java#L767), it will be empty when this call is made. 

Would that create any issues? Since it will not re-issue tokens I think

Reply:
I'm trying to make sure that I understand this correctly. The code link you pasted points to the `getStoppingDatastreamGroupInstances` method, which is used by the leader `Coordinator` to get the hosts that are processing tasks for a stopping stream.

The force stop that I added is supposed to be called only if the stream is stuck in STOPPING state indefinitely. Here's a snapshot of the cluster in that bad state:

Leader did a best effort to stop the stream. It issued tokens for the instances in the `liveTaskMap`, it then updated the `liveTaskMap` (in case of a stopping stream that would amount to removing the stream's tasks from the map). Leader polled the followers and waited for them to acknowledge that they stopped their tasks, and it didn't hear from at least one host. It revokes the said hosts' tokens, alerts that "something went wrong with this host" and refuses to initiate a STOPPING -> STOPPED transition.

At this stage all the token issue/claim/revoke business is done. If someone is calling force stop on a stream, the intent is not to force reissue tokens. It is to reset the state of the stream, cleanup leftover garbage in ZK, and let subsequent operations such as resume go through.

Reply:
Discussed offline. @shrinandthakkar's question was whether we need the old call to `_store.datastreamUpdate` that notifies the leader to initiate an assignment change. While it's true that second and subsequent calls to `stop()` without the force flag won't reissue new tokens and the leader won't have a way of getting feedback from followers, we can't simply remove/modify this call. This call is still useful in case the assignment tokens feature is disabled. Modifying this code path may introduce regressions.